presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -937,7 +937,7 @@ public String toString()
 
     private <T> KeyAndContext<T> getCachingKey(MetastoreContext context, T key)
     {
-        MetastoreContext metastoreContext = metastoreImpersonationEnabled ? context : new MetastoreContext(NO_IMPERSONATION_USER);
+        MetastoreContext metastoreContext = metastoreImpersonationEnabled ? context : new MetastoreContext(NO_IMPERSONATION_USER, context.getQueryId(), context.getClientInfo(), context.getSource());
```

Comment:
Will changing `MetastoreContext` affect how the impersonation stuff works? Looks like this will make the original cache key not working, especially after including query id. cc @arhimondr @BlueStalker

Reply:
The hashcode implementation in `KeyAndContext` uses only context.username, key. So should be fine right ?

Reply:
The caching makes it very complicated.

I think we should always pass original values to the metastore when loading values, so the underlying metastore implementation should see the actual values, and not the stubs from the caching hive metastore. Unfortunately this approach breaks security when using impersonation. When impersonation is enabled we want to have 1 cache entry per user, to make sure the permission checks are properly applied.

I would recommend changing the code a little bit. Let's always pass the original values, but at the same time exclude them from the `equals` and `hashcode` checks (that can be very confusing for the future readers, so please document it in the code). When impersonation is enabled we can include only the username in the `equals` and `hashcode` to make sure 1 entry per user is created. 

What do you think guys?

Reply:
@arhimondr I think changing how cache works is beyond the scope of this PR 🤔. My understanding is that the purpose of this PR is to add new fields in to `MetastoreContext` while not impacting other already-working features. 

Per-user caching could be introduced in a separate PR IMO, in order to make impersonation secure. WDYT?

Reply:
I think the per-user caching logic is already there. To NOT break the logic, i think @arhimondr is suggesting to exclude queryId, other fields in `equals` and `hashcode` of `KeyAndContext`.